base/loading.jl

```diff
@@ -2104,11 +2105,37 @@ function compilecache(pkg::PkgId, path::String, internal_stderr::IO = stderr, in
                 end
             end
 
-            # this is atomic according to POSIX (not Win32):
-            rename(tmppath, cachefile; force=true)
             if cache_objects
-                rename(tmppath_so, ocachefile::String; force=true)
+                try
+                    rename(tmppath_so, ocachefile::String; force=true)
+                catch e
+                    e isa IOError || rethrow()
+                    # Windows prevents renaming a file that is in use so if there is a Julia session started
+                    # with a package image loaded, we cannot rename that file.
+                    # The code belows append a `_i` to the name of the cache file where `i` is the smallest number such that
+                    # that cache file does not exist.
+                    ocachename, ocacheext = splitext(ocachefile)
+                    old_cachefiles = filter(x->startswith(x, basename(ocachename)) && endswith(x, ocacheext), readdir(cachepath))
+                    nums = Set(1:length(old_cachefiles))
+                    for file in old_cachefiles
+                        name = splitext(file)[1]
+                        s = split(name, '_')
+                        if length(s) == 3 # e.g. lLvWP_g5TNZ_3
+                            i = tryparse(Int, last(s))
+                            if i !== nothing
+                                delete!(nums, i)
+                            end
+                        end
+                    end
+                    # TODO: Risk for a race here if some other process grabs this name before us
+                    num = isempty(nums) ? 1 : minimum(nums)
+                    ocachefile = ocachename * "_$num" * ocacheext
```

Comment:
```suggestion
                    old_cachefiles = Set(readdir(cachepath))
                    num = 1
                    while true
                        ocachefile = ocachename * "_$num" * ocacheext
                        # ispath(ocachefile) || break
                        in(ocachefile, old_cachefiles) || break
                    end
                    # TODO: Risk for a race here if some other process grabs this name before us
```

Reply:
Yeah, this part felt a bit convoluted.

Reply:
We could also try to use tempname, since it implements mostly the same logic as this, but possibly does not support suffixes